### ISSUE TITLE
Create New Spans In gRPC Server RPC Methods

### ISSUE THREAD

**Issue by chrisalexander55** [GoogleCloudPlatform/cloud-trace-nodejs#656]
Hi Everyone,

I am trying to create additional trace spans inside my gRPC rpc methods. I reviewed the API docs and examples but have not seen any firm guidance or working implementation. When I do call trace#createChildSpan, a null is returned. What am I doing wrong here? Here is an abbreviated code block from my server.js file:

`const traceApi = require("@google-cloud/trace-agent").start({
  projectId: "festivals-183900",
  keyFilename: googleSvcAccount
});
server.addService(smtpTemplateProto.templatesPackage.SmtpTemplateSvc.service, {
addBindable: (call: any, cb: any) => {
    let clientCn: Client;
    let data: Object;
    let rollBack: boolean = false;
    let spanAddBindable: any;
    try {
      // create spanAddBindable span
      spanAddBindable = traceApi.createChildSpan({
        name: "Svc.BindableController#addBindable"
      });
      console.log(spanAddBindable);   <== evals as null!
      spanAddBindable.endSpan();
  }
  catch(err) {}
}
}
`

**Comment by kjin** [comment]
I think I was able to reproduce this with Node 8 + `async_hooks` tracing, and gRPC 1.6 (I don't expect the gRPC version to matter, though). Would you be able to provide the following details?

* Node version, and whether `GCLOUD_TRACE_NEW_CONTEXT` is set as per [notes on tracing for async/await](https://github.com/GoogleCloudPlatform/cloud-trace-nodejs#tracing-with-asyncawait)
* Any warning messages that pop up when the Trace Agent is started with config option `logLevel: 1` (WARN)? If `GCLOUD_TRACE_NEW_CONTEXT` is set, I would expect to see:

```
WARN:@google-cloud/trace-agent: grpc: Cannot create nested root spans.
```

**Comment by chrisalexander55** [comment]
Sure, give me a few hours to get back to my PC.

**Comment by chrisalexander55** [comment]
Node: 8.8.1
@google-cloud/trace-agent: 2.3.3

I did not set GCLOUD_TRACE_NEW_CONTEXT=1 when executing my node app. I will sit down with this tomorrow night when I can dedicate real time to validate your emerging fix to methods utilizing async/await, which I heavily leverage. Thanks!

**Comment by kjin** [comment]
Ok, thanks for the info! If you're using untranspiled async/await (or rather, since you're using TypeScript, targeting `es2017`) then setting `GCLOUD_TRACE_NEW_CONTEXT` should help (we are looking for feedback, so any reports of how well it works would be extremely helpful). That being said, my initial repro already had `GCLOUD_TRACE_NEW_CONTEXT` set, so the bug it probably something else. I'll see if I can find a different repro.

**Comment by mrene** [comment]
I have a similar issue (node 8.9.4, async/await with `GCLOUD_TRACE_NEW_CONTEXT`, grpc 1.7.3 server) - I get valid traces for the first call and then I start seeing `grpc: Cannot create nested root spans.`
I then added some more tracing inside my own methods and I started seeing `Custom Span API: creating child for an already closed span ... grpc:/...`. After a while, when the traces get pushed, I get another valid trace then the same errors repeat themselves.

The errors relating to the custom spans are reported because `cls.getRootContext()` returns a span that is closed (the span from the last call). Debugging this into `cls-ah` revealed some weird behaviour between `grpc-native-core` and `async_hooks`. Every root span gets created while both the `executionAsyncId` and `triggerAsyncId` are 0; the `before` callback also does not execute before handling a new call and therefore the `context` variable within `cls-ah` points to the last used context.

cls-ah seems to do the following:
- maintain a global context object which is used as cls
- maintain a hash of context objects, indexed by `executionAsyncId`
- whenever an `init` event occurs:
    - copy the current global context object into the hash (regardless of the parent id specified by the callback)
- whenever an `enter` event occurs:
    - if the hash contains the entered id, set the global context object to its content


In this scenario, because we're lacking an `enter`, the first grpc call trace is valid because the context object is initially initialized to `{}` and we process to creating a root span because the context doesn't contain any previous root span. Upon the 2nd call, the context still contains the previous root span (in a closed state), only errors arise. When the traces get flushed, an `enter` event occurs and sets the global `context` object back to its initial value (because the trace writer context was created before the first grpc call, so it copied the initial value). 

At this point this looks more like a grpc bug although it could either be handled in the plugin (using the `AsyncResource` API or within `grpc-native-core` - I'm sure there's a param to carry over execution context ids when switching from native->js).

**Comment by mrene** [comment]
It seems the `AsyncHooks` embedder's api hasn't landed in NAN yet (nodejs/node#13254) so a JS fix is the only possible option.

I've got a hacked up version of the plugin to validate the entire hypothesis and it patches the problem. 

I'd like to get some of your opinions before opening a PR, got some comments/questions here:

- Copying the current context from `init` might not be the right thing, since `init` can happen on its own without `enter`
    - Part of the problem is that contexts created after the first span directly inherit it even if they don't share a parent context
    - Every async context will probably point to the same object 
        - `context` is set back to its original value in `runAndReturn` but as soon as more get created, they will be set back (through `before` and `init`) and the original value will not be restored. 
        - This could be mitigated by implementing `after()` and resetting values accordingly.
- This feels like it `emitDestroy()` could be called by `span#endSpan()` since it basically knows whenever it's done

What do you think?

**Comment by kjin** [comment]
@mrene Thanks for investigating the `async_hooks`-based tracing issue! Please go ahead and open the PR.

I've been thinking through the observations you had, here are my thoughts so far:

> Copying the current context from `init` might not be the right thing, since `init` can happen on its own without `enter`

Copying the current context in `init` is necessary to preserve it across asynchronous boundaries. Maybe I'm misunderstanding what you're saying?

> Part of the problem is that contexts created after the first span directly inherit it even if they don't share a parent context
> Every async context will probably point to the same object 

I think you were correct in creating a new `AsyncResource` for each request being handled, so that subsequent requests after the first one don't point to the same object.

> `context` is set back to its original value in `runAndReturn` but as soon as more get created, they will be set back (through `before` and `init`) and the original value will not be restored. 
> This could be mitigated by implementing `after()` and resetting values accordingly.

I believe implementing `after()` to reset context would effectively be a no-op, since `before()` will run anyway for the next async resource. Would you be able to further articulate what the potential problem could be here?

> This feels like it `emitDestroy()` could be called by `span#endSpan()` since it basically knows whenever it's done

I'm open to (and for this and other reasons, actively thinking about) the idea of creating our own `AsyncResource` object for root spans and emitting the lifecycle events correspondingly. Let's talk about it in a separate issue (please feel free to open it if I haven't already).

> `runAndReturn()` should probably set `contexts[uid]` as well as `context`

What should `contexts[uid]` be set to?

**Comment by kjin** [comment]
@mrene After looking into it more, I believe your reasoning is correct that `after` should be implemented to reset the previous context.

I'll open a PR to implement `after` soon-ish, probably tomorrow, and gRPC server tracing should just work (with NEW_CONTEXT-based tracing, at least) without need for additional changes to the plugin. That being said, it might still make sense for you to open a PR against gRPC to add async_hooks embedder support. (if you feel motivated; I can do it as well 😄 )

**Comment by mrene** [comment]
> Copying the current context in init is necessary to preserve it across asynchronous boundaries. Maybe I'm misunderstanding what you're saying?

I mean `init` shouldn't be constructed as `before` - the runtime isn't switching to it yet. one of its argument is the `triggerAsyncId`. I agree that the context should be copied, just that it should be copied from `contexts[triggerAsyncId]` rather than the `context` global var being set at that moment. (In most cases those will be the same but it's possible to pass another parent in the async_hooks API)

> What should contexts[uid] be set to?

The same thing, basically `context = contexts[uid] = {};` whenever it is being initialized. This would make sure that the context is preserved if an `enter` event to the same context occurs afterwards. It'll give that context its own object, which it doesn't have now (only context initialized while this one is active will be correctly linked). 

Thinking about this more makes me realize that running this without creating another `AsyncResource` could cause the same id to be re-used twice, so it might not be advisable until this is implemented.

I noticed a small difference in behaviour with `continuation-local-storage`, it does not store any data unless the current context has been _activated_, whereas `cls-ah` initializes an empty object so things can technically be saved/fetched even if `runAndReturn` hasn't been called. Nothing critical but it might hide subtle bugs.

**Comment by chrisalexander55** [comment]
@kjin, Sorry I'm getting back to this a day later... I execute my node.js app using the README.md reference above. Unfortunately, it didn't work and generated the following stdout:

(node:17037) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): TypeError: Cannot read property 'endSpan' of null
(node:17037) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that arenot handled will terminate the Node.js process with a non-zero exit code.
(node:17037) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 2): Error: Release called on client which has already been released to the pool.

Based on the evolution of this thread, looks like you guys are on it. If I can lend any fix validation, please hit me up as I would be more than happy to assist. Thanks!

**Comment by kjin** [comment]
@chrisalexander55 Sorry for the late reply. I think what you and @mrene are experiencing are separate issues, as you were originally not using `GCLOUD_TRACE_NEW_CONTEXT`, and the earlier discussion concerned the new behavior associated with it.

It's possible for `createChildSpan` to return `null` even if gRPC server tracing is working correctly; this might happen if you are receiving requests at a rate greater than 10/sec (by default, we sample 10 traces per second). Therefore you'd have to wrap the `span.endSpan` call in a `if (span) {}` block. (This was a conscious design choice at the time the API was introduced, and something I'm now highly considering changing given how brittle it makes user code).

Though it still might be possible that this is a gRPC server tracing bug; however I can't know this unless you have turned on `logLevel: 1` set in the config and watch for warning messages. If you already did, and there weren't more than a few at the beginning, then chances are what I said in the previous paragraph is the culprit, rather than gRPC server tracing.

**Comment by chrisalexander55** [comment]
@kjin, not a problem - we all have multiple items on our plates. I did add `GCLOUD_TRACE_NEW_CONTEXT=1` to my node execution and set trace config `logLevel: 1`, still observing the `null`. I will take your suggestion and add the `if (span) {}` and report the result this evening EST.

**Comment by kjin** [comment]
@chrisalexander55 Thanks!

@mrene I've been doing some thinking about this but I don't want to confound this with the null span issue. I'm going to move our discussion to a separate tracking issue as they're fairly distinct.

**Comment by ofrobots** [comment]
@chrisalexander55 were you able to try out `if (span) {}`?

**Comment by chrisalexander55** [comment]
Out of town and no laptop until Thursday - argh!!

**Comment by chrisalexander55** [comment]
Within a RPC method body, I can successfully create a span but calling span#endSpan AFTER invoking an await method call, the span throws with `TypeError: Cannot read property \'endSpan\' of null`. So it continues to be a defect, at least in my hands. I'm open to any guidance you can provide.

**Comment by kjin** [comment]
@chrisalexander55 Thanks for the update! Do you have sample code that you can provide?

**Comment by chrisalexander55** [comment]
` 
// import { traceApi } from "../server";

  public async addCampaign(call: any): Promise<Object> {
    // 1. type lookup exist?
    // const spanDoesLookupUuidExist = traceApi.createChildSpan({
    //   name: "LookupController#doesLookupUuidExist"
    // });
    const boolMsg1 = await this._lookupCntrl.doesLookupUuidExist(call.request.typeUuid);
    // spanDoesLookupUuidExist.endSpan();

    // handled error msg
    if (boolMsg1 instanceof Msg) {
      // UuidMsg
      return {
        uuid: undefined,
        msg: ToGrpc.from_Msg(boolMsg1)
      };
    }

    if (!boolMsg1) {
      // No
      const msg = new Msg();
      msg.code = BizLogicCode;
      msg.data = undefined;
      msg.text = "Campaign type not found.";

      // UuidMsg
      return {
        uuid: undefined,
        msg: ToGrpc.from_Msg(msg)
      };
    }

    // 2. does category exist?
    // const spanDoesCategoryExistByUuid = traceApi.createChildSpan({
    //   name: "CategoryController#doesCategoryExistByUuid"
    // });
    const boolMsg2 = await this._categoryCntrl.doesCategoryExistByUuid(call.request.categoryUuid);
    // spanDoesCategoryExistByUuid.endSpan();

    // handled error msg
    if (boolMsg2 instanceof Msg) {
      // UuidMsg
      return {
        uuid: undefined,
        msg: ToGrpc.from_Msg(boolMsg2)
      };
    }

    if (!boolMsg2) {
      // No
      const msg = new Msg();
      msg.code = BizLogicCode;
      msg.data = undefined;
      msg.text = "Category not found.";

      // UuidMsg
      return {
        uuid: undefined,
        msg: ToGrpc.from_Msg(msg)
      };
    }

    // 3. does campaign name already exist?
    // const spanDoesCampaignNameExist = traceApi.createChildSpan({
    //   name: "CampaignController#doesCampaignNameExist"
    // });
    const boolMsg3 = await this._campaignCntrl.doesCampaignNameExist(
      call.request.appId,
      call.request.categoryUuid,
      call.request.templateUuid,
      call.request.typeUuid,
      call.request.campaign.name
    );
    // spanDoesCampaignNameExist.endSpan();

    // handled error msg
    if (boolMsg3 instanceof Msg) {
      // UuidMsg
      return {
        uuid: undefined,
        msg: ToGrpc.from_Msg(boolMsg3)
      };
    }

    if (boolMsg3) {
      // Yes
      const msg = new Msg();
      msg.code = BizLogicCode;
      msg.data = undefined;
      msg.text = "Campaign name already exists.";

      // UuidMsg
      return {
        uuid: undefined,
        msg: ToGrpc.from_Msg(msg)
      };
    }

    // 4. persist campaign
    const campaign = new Campaign();
    campaign.name = call.request.campaign.name;
    campaign.description = call.request.campaign.description;

    // const spanAddCampaign = traceApi.createChildSpan({
    //   name: "CampaignController#addCampaign"
    // });
    const uuidMsg = await this._campaignCntrl.addCampaign(
      call.request.appId,
      call.request.categoryUuid,
      call.request.templateUuid,
      call.request.typeUuid,
      campaign,
      call.request.createdBy
    );
    // spanAddCampaign.endSpan();

    // handled error msg
    if (uuidMsg instanceof Msg) {
      // UuidMsg
      return {
        uuid: undefined,
        msg: ToGrpc.from_Msg(uuidMsg)
      };
    }

    // UuidMsg
    return {
      uuid: uuidMsg,
      msg: undefined
    };
  }`

**Comment by kjin** [comment]
Try `if (span) { span.endSpan(); }` or `span && span.endSpan()` -- since `span` can be null due to sampling. (If it still doesn't work for you let us know.)

Given the verbosity of doing so, I'm going to use this as a tracking bug to change behavior so that the Trace Agent always returns non-null spans.

**Comment by chrisalexander55** [comment]
Thanks Kjin. Please let me know when the non-null span implementation is complete as I really don't want to pollute my code with conditional statements that only be necessary for a few months.

**Comment by kjin** [comment]
Hi @chrisalexander55 -- I've released version 2.6 of the Trace Agent, in which spans are now guaranteed to be non-null ([version notes](https://github.com/GoogleCloudPlatform/cloud-trace-nodejs/blob/master/CHANGELOG.md#notable-changes)). Please give it a try and let us know if there are any issues!

**Comment by chrisalexander55** [comment]
Kevin,

Thank you for the notification. I will pull down the bits and kick the
tires this week. All the best!
